python/GafferUI/ColorChooser.py

```diff
@@ -106,15 +108,148 @@ def _displayTransformChanged( self ) :
 		GafferUI.Slider._displayTransformChanged( self )
 		self._qtWidget().update()
 
+class _ColorFieldWidget( QtWidgets.QWidget ) :
+
+	def __init__( self, parent = None ) :
+
+		QtWidgets.QWidget.__init__( self, parent )
+		self.__size = 194 # The height of the sliders and swatches in a `ColorChooser` dialogue.
+
+	def resizeEvent( self, event ) :
+
+		w = event.size().width()
+		h = event.size().height()
+
+		if w != h :
+			self.__size = h
+			self.setMinimumWidth( h )
+
+	def sizeHint( self ) :
+
+		# This is needed to set our size for layout when a dialog using `ColorChooser` is first
+		# opened. Without it, the other widgets in our row would be shrunk considerably.
+
+		return QtCore.QSize( self.__size, self.__size )
+
+class _ColorField( GafferUI.Widget ) :
+
+	def __init__( self, color = imath.Color3f( 1.0 ), staticComponent = "h", **kw ) :
+
+		GafferUI.Widget.__init__( self, _ColorFieldWidget(), **kw )
+
+		self._qtWidget().setMinimumSize( 50, 50 )
+		self._qtWidget().setSizePolicy( QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Expanding )
+
+		self._qtWidget().paintEvent = Gaffer.WeakMethod( self.__paintEvent )
+
+		self.__color = color
+		self.__staticComponent = staticComponent
+		self.__colorFieldToDraw = None
+		self.setStaticComponent( staticComponent, color )
+
+	# Sets the component that is independent of the field. `equivalentColor` is the color
+	# to use for the field in the space of the new static component, which may be different
+	# from the current color space, but should have the same effective color.
+	# `equivalentColor` is in RGB space if the new component is R, G or B, and HSV space
+	# if the new component is H, S or V.
```

Comment:
I don't really understand this `equivalentColor` business - could you explain why it is necessary?

Reply:
It's kind of a weird one, perhaps there's a better solution.

Backing up a bit, we set the color with `setColor()` and the color is stored as-is with no conversion between RGB and HSV space. Later when we make use of the color, it's taken to be in RGB space if the `__staticComponent` is R, G or B, and HSV if `__staticComponent` is H, S or V.

That was motivated by the need to eliminate conversions between the spaces and the precision loss that would entail.

Now when changing `__staticComponent` using `setStaticComponent()`, we may need to change the internally stored color as well. If we're switching from R, G or B `__staticComponent` to H, S or V `__staticComponent` (or the reverse), we need to also change our internal `__color` value to be in the new space. For the same precision loss reason above, I didn't want to just do a conversion, I wanted to get the new color from the caller, who has the RGB- and HSV-spaced colors stored separately.

So `equivalentColor` is the color that should be used for the field for the new static component being set. It may be the same as the current color if you're just switching static components amongst RGB or HSV triplets. Or it may be the same effective color but in a different space and with different numeric values.

Reply:
I wondered if it might have been something like that. Perhaps it would make sense to just have one method, `setColor( color, staticComponent )`? The color and the component are intrinsically tied together because the color space dictates the valid components and the component dictates the required color space. So requiring them to always be provided as a pair makes that a bit clearer perhaps?

Reply:
That sounds like a good improvement, and makes it all clearer too. I also made `getColor()` return at tuple of the color and color space based on that reasoning. It results in a number of cases where the color space is not used, but since we're setting both, it seems best to also get both to keep them symmetric?

Reply:
> I also made getColor() return at tuple of the color and color space...it seems best to also get both to keep them symmetric?

Yep, good call.